pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+    pub type ActiveCollators<T: Config> = StorageMap<
+        _,
+        Twox64Concat,
+        SessionIndex,
+        BoundedBTreeSet<T::CollatorId, T::MaxCollatorsPerSession>,
+        ValueQuery,
+    >;
+
+    /// A list of active collators for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveCollatorsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedBTreeSet<T::CollatorId, T::MaxCollatorsPerSession>, ValueQuery>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Event emitted when the activity tracking status is updated
+        ActivityTrackingStatusSet { status: ActivityTrackingStatus },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The size of a collator set for a session has already reached MaxCollatorsPerSession value
+        MaxCollatorsPerSessionReached,
+        /// Error returned when the activity tracking status is attempted to be updated before the end session
+        ActivityStatusUpdateSuspended,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::set_inactivity_tracking_status())]
+        pub fn set_inactivity_tracking_status(
+            origin: OriginFor<T>,
+            is_enabled: bool,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            let current_status_end_session_index = match <CurrentActivityTrackingStatus<T>>::get() {
+                ActivityTrackingStatus::Enabled { start: _, end } => end,
+                ActivityTrackingStatus::Disabled { end } => end,
+            };
+            let current_session_index = T::CurrentSessionIndex::session_index();
+            ensure!(
+                current_session_index > current_status_end_session_index,
+                Error::<T>::ActivityStatusUpdateSuspended
+            );
+            let new_status_end_session_index =
+                current_session_index + T::MaxInactiveSessions::get();
+            let new_status = if is_enabled {
+                ActivityTrackingStatus::Enabled {
+                    start: current_session_index + 1,
+                    end: new_status_end_session_index,
+                }
+            } else {
+                <ActiveCollatorsForCurrentSession<T>>::put(BoundedBTreeSet::new());
+                ActivityTrackingStatus::Disabled {
+                    end: new_status_end_session_index,
+                }
+            };
+            <CurrentActivityTrackingStatus<T>>::put(new_status.clone());
+            Self::deposit_event(Event::<T>::ActivityTrackingStatusSet { status: new_status });
+            Ok(())
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
```

Comment:
I count 3 reads at least (getSelfChain author, activity ttracking status, and session index

Reply:
Would it make sense to increment the  `total_weight ` before each if instead?

Reply:
yes makes sense